File: Linked list/palindrome_linkedList.py
"""Given a singly linked list, determine if it is a palindrome."""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        l = []
        if head is None:
            return True
        if head.next is None:
            return True
        if head.next.next is None:
            if head.val == head.next.val:
                return True
            else:
                return False
        slow = head
        fast = head
        cnt = 0
        while head is not None:
            cnt += 1
            head =  head.next
        while slow and fast and fast.next:
            l.append(slow.val)
            slow = slow.next
            fast = fast.next.next
        if cnt % 2 == 1:
            slow  = slow.next
        for i in range(len(l)-1,-1,-1):
            if l[i] != slow.val:
                return False
            slow = slow.next
        return True

            
        

File: Linked list/PalindromeLinkedList.cpp
//Problem Link:- https://leetcode.com/problems/palindrome-linked-list/

#include<bits/stdc++.h>
using namespace std;


//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    
    ListNode* helper(ListNode* head){
        ListNode* curr = head;
        ListNode* prev = NULL;
        ListNode* nxt;
        while(curr != NULL){
            nxt = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
        }
        return prev;
    }
    
    bool isPalindrome(ListNode* head) {
        ListNode* node = NULL;
        node = new ListNode(head->val);
        ListNode* temp = head;
        head = head->next;
        ListNode* start = node;
        while (head != NULL){
            node->next = new ListNode(head->val);
            head = head->next;
            node = node->next;
        }
        head = helper(temp);
        while (head != NULL && start != NULL){
            if (head->val != start->val){
                return false;
            }
            head = head->next;
            start = start->next;
        }
        return true;
    }
};